storage: improve the migration away from txn.DeprecatedOrigTimestamp

andreimatei opened the pull request:
19.2 doesn't generally set txn.ReadTimestamp. Instead, it sets
txn.DeprecatedOrigTimestamp. Before this patch, all code dealing with
txn.ReadTimestamp had to deal with the possibility of it not being set.
This is fragile; I recently forgot to deal with it in a patch.
This patch sets txn.ReadTimestamp to txn.DeprecatedOrigTimestamp when it
wasn't set, thereby releaving most other code of they worry.

This comes at the cost of an extra txn clone for requests coming from
19.2 nodes.

Release note: None

andreimatei committed:
storage: improve the migration away from txn.DeprecatedOrigTimestamp

19.2 doesn't generally set txn.ReadTimestamp. Instead, it sets
txn.DeprecatedOrigTimestamp. Before this patch, all code dealing with
txn.ReadTimestamp had to deal with the possibility of it not being set.
This is fragile; I recently forgot to deal with it in a patch.
This patch sets txn.ReadTimestamp to txn.DeprecatedOrigTimestamp when it
wasn't set, thereby releaving most other code of they worry.

This comes at the cost of an extra txn clone for requests coming from
19.2 nodes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44407)
<!-- Reviewable:end -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44407#-:-LzddDcjEaXBL95CGnZE:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

andreimatei committed:
kv: deal with leaf txns created on behalf of 19.2 nodes

Before this patch, a leaf txn created on a 20.1 node on behalf of a 19.2
root didn't have its ReadTimestamp set (it only had
DeprecatedOrigTimestamp). I don't know if this is causing any problems,
but at the very least it's confusing as hell. It's also in my way for
the future.

This patch sets ReadTimestamp when the leaf TxnCoordSender is created.

Release note: None

andreimatei reviewed (commented):
prepended a new commit. PTAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44407#-:-LzmpCo52OpRCZjicqSh:b98zehb)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44407#-:-Lzo9IWX50Vk1NN9rcn8:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/txn_coord_sender.go, line 311 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44407#-Lzo94YQ8_P80sCVkrnX:-Lzo94YRA8gauyzUa-IQ:b-mtx6tx) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d5f521930bb924b7d0d82180d2e0e5f72ed5d94/pkg/kv/txn_coord_sender.go#L311)):*
> ```Go
> 	tis.Txn.AssertInitialized(context.TODO())
> 	// Deal with requests from 19.2 nodes which did not set ReadTimestamp.
> 	txn := &tis.Txn
> ```

Might as well use this throughout this method.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44407#-:-LzrHRKM1yvIoaj5ouDr:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @nvanbenschoten)

___
*[pkg/kv/txn_coord_sender.go, line 311 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44407#-Lzo94YQ8_P80sCVkrnX:-LzrGXDc0oPrVzfX-WJt:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d5f521930bb924b7d0d82180d2e0e5f72ed5d94/pkg/kv/txn_coord_sender.go#L311)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Might as well use this throughout this method.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->
